File: sdk/node/examples/cognitive-loop.ts
import { MuninnClient } from '../src/index';

async function main() {
  const client = new MuninnClient({
    baseUrl: 'http://localhost:8476',
    token: process.env.MUNINN_TOKEN || '',
  });

  try {
    console.log('=== Cognitive Loop Example ===\n');

    const memories = await client.writeBatch('default', [
      {
        concept: 'transformer architecture',
        content: 'Self-attention mechanism allows parallel processing of sequence elements. O(n²) complexity for sequence length n.',
        tags: ['ml', 'architecture'],
      },
      {
        concept: 'attention mechanism',
        content: 'Query-Key-Value attention: softmax(QK^T/√d)V. Multi-head attention uses h parallel attention functions.',
        tags: ['ml', 'attention'],
      },
      {
        concept: 'positional encoding',
        content: 'Sinusoidal position embeddings added to input embeddings. Learned positional embeddings also effective.',
        tags: ['ml', 'encoding'],
      },
    ]);
    const ids = memories.results.map(r => r.id).filter((id): id is string => !!id);
    console.log(`Stored ${ids.length} memories`);

    if (ids.length >= 3) {
      await client.link({ source_id: ids[0], target_id: ids[1], rel_type: 'related', weight: 0.9 });
      await client.link({ source_id: ids[1], target_id: ids[2], rel_type: 'related', weight: 0.8 });
      console.log('Created association graph');
    }

    const result = await client.activate({
      context: ['how does self-attention work in transformers?'],
      limit: 10,
      include_why: true,
    });
    console.log(`\nRecall returned ${result.total_found} memories:`);
    for (const item of result.activations) {
      console.log(`  [${item.score.toFixed(3)}] ${item.concept}`);
    }

    if (ids[0]) {
      const graph = await client.traverse({
        start_id: ids[0],
        max_hops: 2,
        max_nodes: 10,
      });
      console.log(`\nGraph traversal from "${ids[0]}":`);
      console.log(`  Nodes: ${graph.nodes.length}, Edges: ${graph.edges.length}`);
      for (const node of graph.nodes) {
        console.log(`  [depth ${node.depth}] ${node.concept}`);
      }
    }

    if (ids[0]) {
      const explanation = await client.explain({
        engram_id: ids[0],
        query: ['self-attention', 'transformer'],
      });
      console.log(`\nScore explanation for "${explanation.concept}":`);
      console.log(`  Final score: ${explanation.final_score.toFixed(4)}`);
      console.log(`  Would return: ${explanation.would_return}`);
      console.log(`  Components: Semantic=${explanation.components.semantic.toFixed(3)}, Recency=${explanation.components.recency.toFixed(3)}, Confidence=${explanation.components.confidence.toFixed(3)}`);
    }

    if (ids[0]) {
      const links = await client.getLinks(ids[0]);
      console.log(`\nAssociations for ${ids[0]}: ${links.length} links`);
    }
  } finally {
    client.close();
  }
}

main().catch(console.error);

File: sdk/python/examples/cognitive_loop.py
"""Full cognitive loop: write → activate → link associations.

This example demonstrates a complete cognitive workflow:
1. Write a cluster of related memories
2. Activate memory with semantic query
3. Create associations between memories
4. Verify results and inspect database state

Run with: python examples/cognitive_loop.py
"""

import asyncio

from muninn import MuninnClient


async def main():
    """Cognitive loop example."""
    async with MuninnClient("http://localhost:8476") as client:
        print("MuninnDB Cognitive Loop Demo\n")

        vault = "cognitive"

        # Write a cluster of related memories about transformers
        memories = [
            (
                "attention mechanism",
                "Transformer attention: query, key, value matrices. Scaled dot-product attention computes weights as softmax(Q·K^T/√d). Foundation of modern NLP.",
                ["transformer", "attention", "ml"],
            ),
            (
                "self-attention",
                "Self-attention allows each position in a sequence to attend to all other positions. Enables parallel processing and long-range dependencies.",
                ["transformer", "attention", "ml"],
            ),
            (
                "positional encoding",
                "Since transformers have no recurrence, positional encodings inject sequence order. Sinusoidal embeddings with alternating sin/cos frequencies.",
                ["transformer", "encoding", "ml"],
            ),
            (
                "feed-forward network",
                "MLP applied to each position separately and identically. Typically: Dense → ReLU → Dense. Provides non-linearity and feature transformation.",
                ["transformer", "architecture", "ml"],
            ),
            (
                "layer normalization",
                "Normalization applied to each attention and FFN sublayer. Stabilizes training and improves gradient flow. Applied before or after sublayer.",
                ["transformer", "training", "ml"],
            ),
        ]

        print(f"Writing {len(memories)} related engrams to vault '{vault}'...\n")
        ids = {}
        for concept, content, tags in memories:
            eid = await client.write(
                vault=vault,
                concept=concept,
                content=content,
                tags=tags,
                confidence=0.95,
                stability=0.8,
            )
            ids[concept] = eid
            print(f"  ✓ {concept}")
            print(f"    → {eid}")

        print(f"\nWrote {len(ids)} engrams\n")

        # Activate to find relevant memories
        print("Activating with query: 'how does attention work in transformers?'\n")
        result = await client.activate(
            vault=vault,
            context=["attention mechanism", "transformers", "sequence modeling"],
            max_results=10,
            threshold=0.0,
        )

        print(f"Found {result.total_found} relevant memories:")
        for i, item in enumerate(result.activations, 1):
            print(f"  {i}. [{item.score:.3f}] {item.concept}")

        # Create associations between related engrams
        print(f"\nCreating associations between engrams...")

        associations = [
            ("attention mechanism", "self-attention"),
            ("self-attention", "positional encoding"),
            ("attention mechanism", "layer normalization"),
            ("feed-forward network", "layer normalization"),
        ]

        for source_concept, target_concept in associations:
            source_id = ids[source_concept]
            target_id = ids[target_concept]
            await client.link(
                source_id=source_id,
                target_id=target_id,
                vault=vault,
                rel_type=5,
                weight=0.9,
            )
            print(f"  ✓ {source_concept} → {target_concept}")

        print("\n" + "=" * 50)

        # Get database stats including coherence
        stats = await client.stats()
        print(f"\nDatabase statistics:")
        print(f"  Total engrams: {stats.engram_count}")
        print(f"  Total vaults: {stats.vault_count}")
        print(f"  Storage: {stats.storage_bytes:,} bytes")

        if stats.coherence and vault in stats.coherence:
            coherence = stats.coherence[vault]
            print(f"\nVault '{vault}' coherence:")
            print(f"  Score: {coherence.score:.3f}")
            print(f"  Total engrams: {coherence.total_engrams}")
            print(f"  Orphan ratio: {coherence.orphan_ratio:.3f}")
            print(f"  Contradiction density: {coherence.contradiction_density:.3f}")
            print(f"  Duplication pressure: {coherence.duplication_pressure:.3f}")

        # Read one engram to show full details
        print(f"\n" + "=" * 50)
        attention_id = ids["attention mechanism"]
        print(f"\nReading engram: attention mechanism")
        engram = await client.read(attention_id, vault=vault)
        print(f"  ID: {engram.id}")
        print(f"  Concept: {engram.concept}")
        print(f"  Confidence: {engram.confidence:.2f}")
        print(f"  Stability: {engram.stability:.2f}")
        print(f"  Tags: {', '.join(engram.tags)}")
        print(f"  Access count: {engram.access_count}")
        print(f"  State: {engram.state}")


if __name__ == "__main__":
    asyncio.run(main())
